File: extension/src/domain/ship-check.ts
import type { Target } from './target';
import type { ExperimentKind } from './experiment';
import type { PassiveCheckKind } from './passive-check';

export type ShipCheckStepKind = ExperimentKind | PassiveCheckKind;

export type ShipCheckStepStatus =
  | 'PENDING'
  | 'RUNNING'
  | 'DONE'
  | 'SKIPPED'
  | 'ERRORED';

export interface ShipCheckStep {
  kind: ShipCheckStepKind;
  runId: string;
  status: ShipCheckStepStatus;
}

export type ReadinessState =
  | 'READY'
  | 'NEEDS_ATTENTION'
  | 'BLOCKED'
  | 'UNKNOWN';

export type SyncState = 'NOT_SYNCED' | 'SYNCING' | 'SYNCED' | 'SYNC_FAILED';

export interface ShipCheckRun {
  shipCheckId: string;
  target: Target;
  steps: ShipCheckStep[];
  createdAt: number;
  completedAt?: number | undefined;
  readiness: ReadinessState;
  syncState?: SyncState | undefined;
  syncedAt?: number | undefined;
  lastSyncError?: string | undefined;
}

File: backend/internal/domain/shipcheck.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

type Readiness string

const (
	ReadinessReady          Readiness = "READY"
	ReadinessNeedsAttention Readiness = "NEEDS_ATTENTION"
	ReadinessBlocked        Readiness = "BLOCKED"
	ReadinessUnknown        Readiness = "UNKNOWN"
)

func (r Readiness) Valid() bool {
	switch r {
	case ReadinessReady, ReadinessNeedsAttention, ReadinessBlocked, ReadinessUnknown:
		return true
	default:
		return false
	}
}

type StepStatus string

const (
	StepDone    StepStatus = "DONE"
	StepErrored StepStatus = "ERRORED"
	StepSkipped StepStatus = "SKIPPED"
)

func (s StepStatus) Valid() bool {
	switch s {
	case StepDone, StepErrored, StepSkipped:
		return true
	default:
		return false
	}
}

type Step struct {
	Kind    string     `json:"kind"`
	Status  StepStatus `json:"status"`
	Ordinal int        `json:"ordinal"`
}

type ShipCheck struct {
	ID                uuid.UUID  `json:"id"`
	ProjectID         uuid.UUID  `json:"projectId"`
	ClientShipCheckID string     `json:"clientShipCheckId"`
	TargetOrigin      string     `json:"targetOrigin"`
	Readiness         Readiness  `json:"readiness"`
	CreatedAt         time.Time  `json:"createdAt"`
	CompletedAt       time.Time  `json:"completedAt"`
	SyncedAt          time.Time  `json:"syncedAt"`
	Steps             []Step     `json:"steps"`
}
